### PR TITLE
fix(v2): fix docs homepage permalink issues

### PR DESCRIPTION
**Pull request by slorber**
Related to issues:

https://github.com/facebook/docusaurus/pull/2896
https://github.com/facebook/docusaurus/pull/2861
https://github.com/facebook/docusaurus/pull/2652
https://github.com/facebook/docusaurus/issues/2879

--- 

Current docs homepage code almost works but introduce workarounds at the "edges" (for example we have to handle if/else logic specific to the homepage in the theme components).

I think we should do the opposite, fix the feature at the core directly in the metadata, so that it works once and for all.

I've been able to simplify the docs homepage code, as well as fixing some issues at the same time (like the "previous" link not pointing correctly to /docs).

The idea is simple:
- We should add a single `/docs` route per version, with `exact: false`, so that it can match both `/docs` and `/docs/*`
- Docs home should have `permalink: /docs/`

This way, we are able to handle the homepage like any regular page, with the same data file and the same render path (`renderRoutes()`).

Having a single parent route is also a bit better for perf, because it permits to avoid unmounting/remounting the docs layout when navigation from `/docs/` to `/docs/nextDoc`.

This enabled me to remove some code introduced in the themes, in both classic/bootstrap. 

Even if it renders correctly for `/docs`, I had to put a trailing / in the permalink (so the canonical docs home is `/docs/` not `/docs`), not easy to fix and probably not a big deal.

### CODE REVIEW

**Comment by slorber** [comment]
Hi @lex111 @fanny , this PR is ready to review. 

I was able to remove a few lines of code and simplify the docs home implementation. The themes should not need to handle the homepage in a specific way anymore.

This also fixes the "previous" link here: https://github.com/facebook/docusaurus/pull/2896#issuecomment-640590389

File: packages/docusaurus-plugin-content-docs/src/index.ts [hunk 1]
```diff
@@ -96,24 +98,6 @@ export default function pluginContentDocs(
...
-      'i',
-    ).exec(permalink);
-
-    if (documentIdMatch) {
-      return documentIdMatch[1] === options.homePageId;
-    }
-
-    return false;
-  };
-
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/index.ts:116]: @lex111 code I was able to delete by using permalink = / in metadatas

File: packages/docusaurus-plugin-content-docs/src/index.ts [hunk 2]
```diff
@@ -361,50 +343,8 @@ Available document ids=
...
-                  versionDocsPathPrefix,
-                ]),
-                component: docLayoutComponent,
-                exact: true,
-                modules: {
-                  docsMetadata: aliasedSource(docsBaseMetadataPath),
-                  content: metadataItem.source,
-                },
-              });
-            }
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/index.ts:406]: @lex111 code I was able to delete by using permalink = / in metadatas

File: packages/docusaurus-theme-bootstrap/src/theme/DocPage/index.js [hunk 3]
```diff
@@ -8,49 +8,36 @@
...
         sidebar={sidebar}
       />
       <section className="offset-1 mr-4 mt-4 col-xl-6 offset-xl-4 p-0 justify-content-center align-self-center overflow-hidden">
         <MDXProvider components={MDXComponents}>
-          {isHomePage ? (
-            <DocItem content={content} />
-          ) : (
-            renderRoutes(baseRoute.routes)
-          )}
+          {renderRoutes(baseRoute.routes)}
```
Comment (slorber) [inline packages/docusaurus-theme-bootstrap/src/theme/DocPage/index.js:40]: @lex111 was able to revert the changes that are no longer required for docs homepage support

Comment (slorber) [reply packages/docusaurus-theme-bootstrap/src/theme/DocPage/index.js:40]: @fanny actually as it's bootstrap :)

File: packages/docusaurus-theme-classic/src/theme/DocPage/index.js [hunk 4]
```diff
@@ -59,19 +45,15 @@ function DocPage(props) {
...
           </div>
         )}
         <main className={styles.docMainContainer}>
           <MDXProvider components={MDXComponents}>
-            {isHomePage ? (
-              <DocItem content={content} />
-            ) : (
-              renderRoutes(baseRoute.routes)
-            )}
+            {renderRoutes(baseRoute.routes)}
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/DocPage/index.js:56]: @lex111 was able to revert the changes that are no longer required for docs homepage support

File: packages/docusaurus-theme-classic/src/theme/DocSidebar/index.js [hunk 5]
```diff
@@ -27,9 +27,15 @@ function usePrevious(value) {
...
+// Compare the 2 paths, ignoring trailing /
+const isSamePath = (path1, path2) => {
+  const normalize = (str) => (str.endsWith('/') ? str : `${str}/`);
+  return normalize(path1) === normalize(path2);
+};
+
 const isActiveSidebarItem = (item, activePath) => {
   if (item.type === 'link') {
-    return item.href === activePath;
+    return isSamePath(item.href, activePath);
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/DocSidebar/index.js:38]: consider /docs is active if path is /docs/

File: packages/docusaurus-plugin-content-docs/src/metadata.ts [hunk 6]
```diff
@@ -101,6 +101,14 @@ export default async function processMetadata({
     throw new Error('Document id cannot include "/".');
   }
 
+  const isDocsHomePage = homePageId === baseID;
+
+  if (frontMatter.slug && isDocsHomePage) {
+    throw new Error(
+      `The docs homepage (homePageId=${homePageId}) is not allowed to have a frontmatter slug=${frontMatter.slug} => you have to chooser either homePageId or slug, not both`,
+    );
+  }
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/metadata.ts:110]: @lex111 introduce a new warning, as discussed in https://github.com/facebook/docusaurus/issues/2879

As we don't duplicate the docs home (ie no /docs + /docs/introduction), then we should rather prevent the user to use the :slug frontmatter, as that wouldn't make sense)

**Comment by slorber** [comment]
> I feel so dumb because I really could go this easier, more natural way to implement the docs homepage feature 🤦

That happens 🤪 

--- 

Thanks, will take a look to fix that.

I'm not sure to understand what you mean by nested docs for now ^^

**Comment by slorber** [comment]
@lex111 

---

I fixed the first issue related to docs in subfolders, deployed in the preview a test:

https://deploy-preview-2905--docusaurus-2.netlify.app/docs/next/

BTW this raises another question that's visible in the: as I selected as home doc id a new doc, and older versions do not have this doc, this basically breaks the homepage of other versions.

The home of current version is now broken:

https://deploy-preview-2905--docusaurus-2.netlify.app/docs/


Shouldn't we version this homeDocId? cc @yangshun 

It looks like I should be able to say:
- current version home is introduction
- next version home is getting-started

---

For the 2nd issue about / as root for docs, that's quite complicated 🤪
It goes deep into the routesChunkName.json file, as 2 routes (docItem/docPage) tries to write the entry for "/", and we end up with a bad merge:

```js
  "/": {
    "component": "component---theme-doc-item-178-a40",
    "docsMetadata": "docsMetadata---ec-1-ddf",
    "content": "content---ec-6-8fc"
  },
```

Before we had:

```js
  "/": {
    "component": "component---theme-doc-page-1-be-9be",
    "docsMetadata": "docsMetadata---c-78-ad7",
    "content": "content---ec-6-8fc"
  },
  "/:route": {
    "component": "component---theme-doc-page-1-be-9be",
    "docsMetadata": "docsMetadata---routeec-1-f92"
  },
```

Will try to find a solution, not so simple :)


BTW the current master isn't working perfectly either:
- incorrect sidebar highlighting
- the homepage works, but the sidebar link to the homepage (installation) does not lead anywhere

**Comment by yangshun** [comment]
> Shouldn't we version this homeDocId? 

Yes I think so, but not sure how to do so or if it's even possible, given that we don't version the docusaurus.config.

**Comment by slorber** [comment]
> Yes I think so, but not sure how to do so or if it's even possible, given that we don't version the docusaurus.config.

Same, but not sure how to do that without introducing a new versioned file 🤪 I think it can be fine for now, it probably does not happen often that the user decides to change the homepage. Likely, The homepage feature is likely the most important for the current version IMHO, it's probably not a big deal if the user does not have the homepage on /next if he does the change. 

One hacky way to solve that is to allow passing an array: `homePageId: ["newNextVersionHomePageId","oldHomePageId"]`. I'd rather not add that but this is a potential hack we can add if users complain.

**Comment by slorber** [comment]
@lex111 @yangshun can you review this please?
